From ad2ac6245bbbbd2ae306e3ea4a6660804069e176 Mon Sep 17 00:00:00 2001
From: Amithash Prasasd <amithash@meta.com>
Date: Thu, 21 Nov 2024 21:00:02 -0800
Subject: [PATCH 05/16] Add structured logging for Threshold events

Add capability to create structured logs for threshold crossing
(compatible with Redfish Sensor Events) along with ability to
resolve the logs when the threshold event recovers.

Tested:
Create a fake external sensor named HostDevTemp with Lower Critical
Threshold as 10 and Upper Critical Threshold as 20.
Note: busctl output was modified to ease readability -- Removed
emits-change and writable

Move sensors below lower threshold and get log. We expect a new log
entry with message ReadingBelowLowerCriticalThreshold and Resolved
as false.
```
root@bmc:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/2 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData                   property  as        8 "READING_VALUE=1.0" "SENSOR_NAME=/xyz/openbmc_project/sensors/temperature/HostDevTemp" "THRESHOLD_VALUE=10.0" "UNITS=xyz.openbmc_project.Sensor.Value.Unit.DegreesC" "_CODE_FILE=<snip>" "_CODE_FUNC=<snip>" "_CODE_LINE=588" "_PID=687"
.Message                          property  s         "xyz.openbmc_project.Sensor.Threshold.ReadingBelowLowerCriticalThreshold"
.Resolved                         property  b         false
```
Move sensor back to normal reading. Expected, first log Resolved set
to true and there is a new event log for SensorReadingNormalRange
```
root@bmc:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/2 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData                   property  as        8 "READING_VALUE=1.0" "SENSOR_NAME=/xyz/openbmc_project/sensors/temperature/HostDevTemp" "THRESHOLD_VALUE=10.0" "UNITS=xyz.openbmc_project.Sensor.Value.Unit.DegreesC" "_CODE_FILE=<snip>" "_CODE_FUNC=<snip>" "_CODE_LINE=588" "_PID=687"
.Message                          property  s         "xyz.openbmc_project.Sensor.Threshold.ReadingBelowLowerCriticalThreshold"
.Resolved                         property  b         true

root@bmc:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/4 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData                   property  as        7 "READING_VALUE=15.0" "SENSOR_NAME=/xyz/openbmc_project/sensors/temperature/HostDevTemp" "UNITS=xyz.openbmc_project.Sensor.Value.Unit.DegreesC" "_CODE_FILE=<snip>" "_CODE_FUNC=<snip>" "_CODE_LINE=500" "_PID=687"
.Message                          property  s         "xyz.openbmc_project.Sensor.Threshold.SensorReadingNormalRange"
.Resolved                         property  b         false
```
Repeat for Upper Critical Threshold.

Change-Id: Icaa5fef672f7570bf0cc2efd9d350261115f169a
Signed-off-by: Amithash Prasasd <amithash@meta.com>
---
 meson.build        |   2 +
 src/Thresholds.cpp | 152 ++++++++++++++++++++++++++++++++++++++++++++-
 src/Thresholds.hpp |  10 +++
 src/sensor.hpp     |  15 +++++
 4 files changed, 178 insertions(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 0110986..c7f6996 100644
--- a/meson.build
+++ b/meson.build
@@ -45,10 +45,12 @@ boost = dependency(
 )
 sdbusplus = dependency('sdbusplus')
 phosphor_logging_dep = dependency('phosphor-logging')
+phosphor_dbus_interfaces_dep = dependency('phosphor-dbus-interfaces')
 
 default_deps = [
     boost,
     nlohmann_json_dep,
+    phosphor_dbus_interfaces_dep,
     phosphor_logging_dep,
     sdbusplus,
     uring,
diff --git a/src/Thresholds.cpp b/src/Thresholds.cpp
index f46fd46..df6e8fb 100644
--- a/src/Thresholds.cpp
+++ b/src/Thresholds.cpp
@@ -8,18 +8,28 @@
 #include <boost/asio/error.hpp>
 #include <boost/asio/steady_timer.hpp>
 #include <boost/container/flat_map.hpp>
+#include <phosphor-logging/commit.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
 #include <sdbusplus/exception.hpp>
 #include <sdbusplus/message.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/Sensor/Threshold/event.hpp>
+#include <xyz/openbmc_project/Sensor/Value/common.hpp>
+#include <xyz/openbmc_project/Sensor/event.hpp>
 
+#include <algorithm>
 #include <array>
 #include <chrono>
 #include <cstddef>
 #include <cstdint>
+#include <exception>
+#include <functional>
 #include <limits>
+#include <map>
 #include <memory>
+#include <stdexcept>
 #include <string>
 #include <tuple>
 #include <utility>
@@ -238,11 +248,13 @@ static int cLoFalse = 0;
 static int cLoMidstate = 0;
 static int cDebugThrottle = 0;
 static constexpr int assertLogCount = 10;
+using Unit = sdbusplus::common::xyz::openbmc_project::sensor::Value::Unit;
 
 struct ChangeParam
 {
     ChangeParam(Threshold whichThreshold, bool status, double value) :
-        threshold(whichThreshold), asserted(status), assertValue(value)
+        threshold(std::move(whichThreshold)), asserted(status),
+        assertValue(value)
     {}
 
     Threshold threshold;
@@ -446,6 +458,136 @@ void checkThresholdsPowerDelay(const std::weak_ptr<Sensor>& weakSensor,
     }
 }
 
+void logDeassertThresholds(Sensor* sensor, double value,
+                           thresholds::Level level,
+                           thresholds::Direction direction)
+{
+    namespace events =
+        sdbusplus::event::xyz::openbmc_project::sensor::Threshold;
+
+    auto thresholdIt =
+        std::find_if(sensor->thresholds.begin(), sensor->thresholds.end(),
+                     [level, direction](auto& th) {
+                         return th.level == level && th.direction == direction;
+                     });
+    if (thresholdIt != sensor->thresholds.end())
+    {
+        auto& threshold = *thresholdIt;
+        if (threshold.assertedLog)
+        {
+            try
+            {
+                lg2::resolve(*threshold.assertedLog);
+            }
+            catch (std::exception& ec)
+            {
+                lg2::error("Unable to resolve {LOG} : {ERROR}", "LOG",
+                           std::string(*threshold.assertedLog), "ERROR",
+                           ec.what());
+            }
+            threshold.assertedLog.reset();
+        }
+    }
+    auto it = std::find_if(
+        sensor->thresholds.begin(), sensor->thresholds.end(),
+        [](auto& th) -> bool { return th.assertedLog.has_value(); });
+    // Return if there are outstanding asserts.
+    if (it != sensor->thresholds.end())
+    {
+        return;
+    }
+    auto objPath = sensor->sensorInterface->get_object_path();
+    lg2::commit(events::SensorReadingNormalRange(
+        "SENSOR_NAME", objPath, "READING_VALUE", value, "UNITS",
+        sensor->units));
+}
+
+template <typename errorObj>
+auto logAssertThresholdHelper(const std::string& objPath, double assertValue,
+                              Unit unit, double thresholdValue)
+    -> sdbusplus::message::object_path
+{
+    return lg2::commit(
+        errorObj("SENSOR_NAME", objPath, "READING_VALUE", assertValue, "UNITS",
+                 unit, "THRESHOLD_VALUE", thresholdValue));
+}
+
+void logAssertThresholds(Sensor* sensor, double assertValue,
+                         thresholds::Level level,
+                         thresholds::Direction direction)
+{
+    namespace errors =
+        sdbusplus::error::xyz::openbmc_project::sensor::Threshold;
+    static const std::map<std::tuple<thresholds::Level, thresholds::Direction>,
+                          std::function<sdbusplus::message::object_path(
+                              const std::string&, double, Unit, double)>>
+        thresholdLogMap = {
+            {{Level::WARNING, Direction::HIGH},
+             &logAssertThresholdHelper<
+                 errors::ReadingAboveUpperWarningThreshold>},
+            {{Level::WARNING, Direction::LOW},
+             &logAssertThresholdHelper<
+                 errors::ReadingBelowLowerWarningThreshold>},
+            {{Level::PERFORMANCELOSS, Direction::HIGH},
+             &logAssertThresholdHelper<
+                 errors::ReadingAboveUpperPerformanceLossThreshold>},
+            {{Level::PERFORMANCELOSS, Direction::LOW},
+             &logAssertThresholdHelper<
+                 errors::ReadingBelowLowerPerformanceLossThreshold>},
+            {{Level::CRITICAL, Direction::HIGH},
+             &logAssertThresholdHelper<
+                 errors::ReadingAboveUpperCriticalThreshold>},
+            {{Level::CRITICAL, Direction::LOW},
+             &logAssertThresholdHelper<
+                 errors::ReadingBelowLowerCriticalThreshold>},
+            {{Level::SOFTSHUTDOWN, Direction::HIGH},
+             &logAssertThresholdHelper<
+                 errors::ReadingAboveUpperSoftShutdownThreshold>},
+            {{Level::SOFTSHUTDOWN, Direction::LOW},
+             &logAssertThresholdHelper<
+                 errors::ReadingBelowLowerSoftShutdownThreshold>},
+            {{Level::HARDSHUTDOWN, Direction::HIGH},
+             &logAssertThresholdHelper<
+                 errors::ReadingAboveUpperHardShutdownThreshold>},
+            {{Level::HARDSHUTDOWN, Direction::LOW},
+             &logAssertThresholdHelper<
+                 errors::ReadingBelowLowerHardShutdownThreshold>}};
+    auto objPath = sensor->sensorInterface->get_object_path();
+    auto thresholdIter =
+        std::find_if(sensor->thresholds.begin(), sensor->thresholds.end(),
+                     [level, direction](auto& th) {
+                         return th.level == level && th.direction == direction;
+                     });
+    if ((thresholdIter == sensor->thresholds.end()))
+    {
+        lg2::error("Unable to find threshold {SENSOR}", "SENSOR", objPath);
+        return;
+    }
+    auto& threshold = *thresholdIter;
+    if (threshold.assertedLog)
+    {
+        // Technically we should never get here. But handle anyway.
+        lg2::error("Ignoring new log with unresolved outstanding entry: {LOG}",
+                   "LOG", std::string(*(threshold.assertedLog)));
+        return;
+    }
+    try
+    {
+        threshold.assertedLog = thresholdLogMap.at({level, direction})(
+            objPath, assertValue, sensor->units, threshold.value);
+    }
+    catch (std::out_of_range& e)
+    {
+        lg2::commit(sdbusplus::error::xyz::openbmc_project::Sensor::
+                        InvalidSensorReading("SENSOR_NAME", objPath));
+    }
+    catch (std::exception& e)
+    {
+        lg2::error("Could not create threshold log entry for {SENSOR}",
+                   "SENSOR", objPath);
+    }
+}
+
 void assertThresholds(Sensor* sensor, double assertValue,
                       thresholds::Level level, thresholds::Direction direction,
                       bool assert)
@@ -482,6 +624,14 @@ void assertThresholds(Sensor* sensor, double assertValue,
             lg2::error(
                 "Failed to send thresholdAsserted signal with assertValue");
         }
+        if (assert)
+        {
+            logAssertThresholds(sensor, assertValue, level, direction);
+        }
+        else
+        {
+            logDeassertThresholds(sensor, assertValue, level, direction);
+        }
     }
 }
 
diff --git a/src/Thresholds.hpp b/src/Thresholds.hpp
index 17c2649..e220b38 100644
--- a/src/Thresholds.hpp
+++ b/src/Thresholds.hpp
@@ -5,6 +5,7 @@
 #include <boost/asio/io_context.hpp>
 #include <boost/asio/steady_timer.hpp>
 #include <sdbusplus/asio/connection.hpp>
+#include <sdbusplus/message/native_types.hpp>
 
 #include <array>
 #include <cstddef>
@@ -12,6 +13,7 @@
 #include <limits>
 #include <list>
 #include <memory>
+#include <optional>
 #include <string>
 #include <utility>
 #include <vector>
@@ -48,6 +50,7 @@ struct Threshold
     double value;
     double hysteresis;
     bool writeable;
+    std::optional<sdbusplus::message::object_path> assertedLog;
 
     bool operator==(const Threshold& rhs) const
     {
@@ -60,6 +63,13 @@ void assertThresholds(Sensor* sensor, double assertValue,
                       thresholds::Level level, thresholds::Direction direction,
                       bool assert);
 
+void logDeassertThresholds(Sensor* sensor, double value,
+                           thresholds::Level level,
+                           thresholds::Direction direction);
+void logAssertThresholds(Sensor* sensor, double assertValue,
+                         thresholds::Level level,
+                         thresholds::Direction direction);
+
 struct TimerUsed
 {
     bool used;
diff --git a/src/sensor.hpp b/src/sensor.hpp
index b21df1e..b5c922e 100644
--- a/src/sensor.hpp
+++ b/src/sensor.hpp
@@ -10,6 +10,8 @@
 #include <sdbusplus/asio/connection.hpp>
 #include <sdbusplus/asio/object_server.hpp>
 #include <sdbusplus/exception.hpp>
+#include <sdbusplus/message/native_types.hpp>
+#include <xyz/openbmc_project/Sensor/Value/common.hpp>
 
 #include <array>
 #include <cerrno>
@@ -19,6 +21,7 @@
 #include <functional>
 #include <limits>
 #include <memory>
+#include <stdexcept>
 #include <string>
 #include <utility>
 #include <vector>
@@ -93,6 +96,10 @@ struct Sensor
     virtual ~Sensor() = default;
     virtual void checkThresholds() = 0;
     std::string name;
+
+    using Unit = sdbusplus::common::xyz::openbmc_project::sensor::Value::Unit;
+    Unit units = Unit::DegreesC;
+
     std::string configurationPath;
     std::string configInterface;
     bool isSensorSettable;
@@ -276,6 +283,14 @@ struct Sensor
             setupPowerMatch(dbusConnection);
         }
 
+        auto unitOpt = sdbusplus::message::convert_from_string<Unit>(unit);
+        if (!unitOpt.has_value())
+        {
+            lg2::error("Invalid unit {UNIT} for sensor {NAME}", "UNIT", unit,
+                       "NAME", name);
+            throw std::invalid_argument("Invalid units for sensor");
+        }
+        units = unitOpt.value();
         createAssociation(association, configurationPath);
 
         sensorInterface->register_property("Unit", unit);
-- 
2.47.3

